server/cache.go

```diff
@@ -1,132 +1,342 @@
+	item := &searchKeyItem{
+		key:    searchKey(encodeRegionEndKey(region)),
+		region: region,
 	}
+	// TODO: if duplicated, panic!!!
+	r.searchRegions.ReplaceOrInsert(item)
+	r.regions[region.GetId()] = region
+}
 
-	region, ok := r.leaderRegions[randRegionID]
-	if ok {
-		return region.clone()
+func (r *RegionsInfo) updataRegion(region *metapb.Region) {
+	item := &searchKeyItem{
+		key:    searchKey(encodeRegionEndKey(region)),
+		region: region,
+	}
+	// TODO: if missing, panic!!!
+	r.searchRegions.ReplaceOrInsert(item)
+	r.regions[region.GetId()] = region
+}
+
+func (r *RegionsInfo) removeRegion(region *metapb.Region) {
+	item := &searchKeyItem{
+		key:    searchKey(encodeRegionEndKey(region)),
+		region: region,
 	}
+	regionID := region.GetId()
 
-	return nil
+	// TODO: if missing, panic!!!
+	r.searchRegions.Delete(item)
+	delete(r.regions, region.GetId())
+
+	r.leaders.remove(regionID)
 }
 
-func (r *RegionsInfo) addRegion(region *metapb.Region, leaderPeer *metapb.Peer) {
-	r.Lock()
-	defer r.Unlock()
+// HeartbeatResp is the response after heartbeat handling.
+// If PutRegion is not nil, we should update it in etcd,
+// if RemoveRegion is not nil, we should remove it in etcd.
+type HeartbeatResp struct {
+	PutRegion    *metapb.Region
+	RemoveRegion *metapb.Region
+}
 
-	regionID := region.GetId()
-	cacheRegion, regionExist := r.leaderRegions[regionID]
-	if regionExist {
-		// If region epoch and leader peer has not been changed, return directly.
-		if cacheRegion.region.GetRegionEpoch().GetVersion() == region.GetRegionEpoch().GetVersion() &&
-			cacheRegion.region.GetRegionEpoch().GetConfVer() == region.GetRegionEpoch().GetConfVer() &&
-			cacheRegion.peer.GetId() == leaderPeer.GetId() {
-			return
-		}
+func (r *RegionsInfo) heartbeatVersion(region *metapb.Region) (bool, *metapb.Region, error) {
+	// For split, we should handle heartbeat carefully.
+	// E.g, for region 1 [a, c) -> 1 [a, b) + 2 [b, c).
+	// after split, region 1 and 2 will do heartbeat independently.
+	startKey := encodeRegionStartKey(region)
+	endKey := encodeRegionEndKey(region)
+
+	searchRegion := r.getRegion(region.GetStartKey())
+	if searchRegion == nil {
+		// Find no range after start key, insert directly.
+		r.addRegion(region)
+		return true, nil, nil
+	}
 
-		// If region leader has been changed, remove old region from store cache.
-		oldLeaderPeer := cacheRegion.peer
-		if oldLeaderPeer.GetId() != leaderPeer.GetId() {
-			storeID := oldLeaderPeer.GetStoreId()
-			storeRegions, storeExist := r.storeLeaderRegions[storeID]
-			if storeExist {
-				delete(storeRegions, regionID)
-				if len(storeRegions) == 0 {
-					delete(r.storeLeaderRegions, storeID)
-				}
-			}
+	searchStartKey := encodeRegionStartKey(searchRegion)
+	searchEndKey := encodeRegionEndKey(searchRegion)
+
+	if startKey == searchStartKey && endKey == searchEndKey {
+		// we are the same, must check epoch here.
+		if err := checkStaleRegion(searchRegion, region); err != nil {
```

Comment:
no need now, because now the mechanism can guarantee that if range is same, the version is same.
But we may handle this later if we introduce merge. @qiuyesuifeng has added a comment here in another PR.